File: 128/128_xjf.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var longestConsecutive = function(nums) {
    if (nums.length == 0) return 0;
    let min = 0, a = [];
    for (let n of nums) min = Math.min(min, n);
    for (let n of nums) a[n - min] = n;
    let last = -NaN, cur = 0, res = 1;
    for (let n in a) {
        n = a[n];
        if (n == last + 1) res = Math.max(res, ++cur);
        else cur = 1;
        last = n;
    }
    return res;
};

File: 128/128_xjf.cpp
class Solution {
public:
    unordered_map<int, int> mp;
    int longestConsecutive(vector<int>& nums) {
        if (nums.size() == 0) return 0;
        for (auto &n : nums) mp[n] = 0;
        int res = 1;
        for (auto &p : mp) res = max(check(p.first), res);
        return res;
    }

    int check(int n) {
        if (mp[n] > 0) return mp[n];
        if (mp.find(n - 1) != mp.end())
            return mp[n] = check(n - 1) + 1;
        return mp[n] = 1;
    }
};